File: src/main/java/org/blacksun/graph/algorithms/GraphPath.java
package org.blacksun.graph.algorithms;

import org.blacksun.graph.channel.Channel;
import org.blacksun.graph.node.GraphNode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Spliterator;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public final class GraphPath implements Iterable<Channel> {
    private final ArrayList<Channel> links;
    private final boolean cycle;

    public GraphPath() {
        this(false);
    }

    public GraphPath(boolean cycle) {
        this.cycle = cycle;
        links = new ArrayList<>();
    }

    public GraphPath add(Channel channel) {
        links.add(channel);
        return this;
    }

    public GraphPath clear() {
        links.clear();
        return this;
    }

    public GraphPath remove() {
        if (!links.isEmpty())
            links.remove(links.size() - 1);
        return this;
    }

    public int getLength() {
        return links.size();
    }

    public int getWeight() {
        return links.stream().mapToInt(Channel::getWeight).sum();
    }

    public boolean exists() {
        return cycle || !links.isEmpty();
    }

    @NotNull
    @Override
    public Iterator<Channel> iterator() {
        return links.iterator();
    }

    @Override
    public void forEach(Consumer<? super Channel> action) {
        links.forEach(action);
    }

    @Override
    public Spliterator<Channel> spliterator() {
        return links.spliterator();
    }

    @Override
    public String toString() {
        if (!exists())
            return "";
        if (links.isEmpty())
            return "Access to self";
        StringBuilder base = new StringBuilder();
        base.append("GraphPath[length=")
                .append(getLength())
                .append(", weight=")
                .append(getWeight())
                .append("]:\n\t")
                .append(links.get(0).getFromNode())
                .append(" --> ");
        return base + links.stream()
                .map(Channel::getToNode)
                .map(Object::toString)
                .collect(Collectors.joining(" --> "));
    }

    public GraphNode getFrom() {
        return links.get(0).getFromNode();
    }

    public GraphNode getTo() {
        return links.get(links.size() - 1).getToNode();
    }

    public boolean ofNodes(GraphNode from, GraphNode to) {
        if (from == to && cycle)
            return true;
        if (links.isEmpty())
            return false;
        return getFrom().equals(from) &&
                getTo().equals(to);
    }

    // sending package in datagram mode
    public void close(int weight) {
        int sum = 0;
        for (int i = links.size() - 1; i >= 0; --i) {
            Channel channel = links.get(i);
            if (weight > sum) {
                sum += channel.getWeight();
            } else {
                channel.setUsed(false);
            }
        }
    }
}

File: src/main/java/org/blacksun/graph/algorithms/GraphPath.kt
package org.blacksun.graph.algorithms

import org.blacksun.graph.channel.Channel
import org.blacksun.graph.node.GraphNode

import java.util.ArrayList
import java.util.Spliterator
import java.util.function.Consumer

class GraphPath @JvmOverloads constructor(private val cycle: Boolean = false, private val links: ArrayList<Channel> = ArrayList()) : Iterable<Channel> {

    val length: Int
        get() = links.size

    val weight: Int
        get() = links.sumBy(Channel::weight)

    val from: GraphNode
        get() = links[0].fromNode

    val to: GraphNode
        get() = links[links.size - 1].toNode

    fun add(channel: Channel): GraphPath {
        links.add(channel)
        return this
    }

    fun clear(): GraphPath {
        links.clear()
        return this
    }

    fun remove(): GraphPath {
        if (!links.isEmpty())
            links.removeAt(links.size - 1)
        return this
    }

    fun exists(): Boolean {
        return cycle || !links.isEmpty()
    }

    override fun iterator(): Iterator<Channel> {
        return links.iterator()
    }

    override fun forEach(action: Consumer<in Channel>) {
        links.forEach(action)
    }

    override fun spliterator(): Spliterator<Channel> {
        return links.spliterator()
    }

    override fun toString(): String {
        if (!exists())
            return ""
        if (links.isEmpty())
            return "Access to self"
        val base = StringBuilder()
        base.append("GraphPath[length=")
                .append(length)
                .append(", weight=")
                .append(weight)
                .append("]:\n\t")
                .append(links[0].fromNode)
                .append(" --> ")
        return base.toString() + links
                .map(Channel::toNode)
                .joinToString(" --> ")
    }

    fun ofNodes(from: GraphNode, to: GraphNode): Boolean {
        if (from === to && cycle)
            return true
        return if (links.isEmpty()) false else from == from && to == to
    }

    // sending package in datagram mode
    fun close(weight: Int): Boolean {
        var sum = 0
        var result = false
        for (i in links.indices.reversed()) {
            val channel = links[i]
            if (weight > sum) {
                sum += channel.weight
            } else {
                channel.isUsed = false
                result = true
            }
        }
        return result
    }
}
